extern/storage-sealing/input.go

```diff
@@ -0,0 +1,412 @@
+
+		if offset.Padded()+padLength+deal.size.Padded() > abi.PaddedPieceSize(ssize) {
+			// todo: this is rather unlikely to happen, but in case it does, return the deal to waiting queue instead of failing it
+			deal.accepted(sector.SectorNumber, offset, xerrors.Errorf("piece %s assigned to sector %d with not enough space", piece, sector.SectorNumber))
+			continue
+		}
+
+		offset += padLength.Unpadded()
+
+		for _, p := range pads {
+			ppi, err := m.sealer.AddPiece(sectorstorage.WithPriority(ctx.Context(), DealSectorPriority),
+				m.minerSector(sector.SectorType, sector.SectorNumber),
+				pieceSizes,
+				p.Unpadded(),
+				NewNullReader(p.Unpadded()))
+			if err != nil {
+				err = xerrors.Errorf("writing padding piece: %w", err)
+				deal.accepted(sector.SectorNumber, offset, err)
+				return ctx.Send(SectorAddPieceFailed{err})
+			}
+
+			pieceSizes = append(pieceSizes, p.Unpadded())
+			res.NewPieces = append(res.NewPieces, Piece{
+				Piece: ppi,
+			})
+		}
+
+		ppi, err := m.sealer.AddPiece(sectorstorage.WithPriority(ctx.Context(), DealSectorPriority),
+			m.minerSector(sector.SectorType, sector.SectorNumber),
+			pieceSizes,
+			deal.size,
+			deal.data)
+		if err != nil {
+			err = xerrors.Errorf("writing piece: %w", err)
+			deal.accepted(sector.SectorNumber, offset, err)
+			return ctx.Send(SectorAddPieceFailed{err})
+		}
+
+		log.Infow("deal added to a sector", "deal", deal.deal.DealID, "sector", sector.SectorNumber, "piece", ppi.PieceCID)
+
+		deal.accepted(sector.SectorNumber, offset, nil)
+
+		offset += deal.size
+		pieceSizes = append(pieceSizes, deal.size)
+
+		res.NewPieces = append(res.NewPieces, Piece{
+			Piece:    ppi,
+			DealInfo: &deal.deal,
+		})
+	}
+
+	return ctx.Send(res)
+}
+
+func (m *Sealing) handleAddPieceFailed(ctx statemachine.Context, sector SectorInfo) error {
+	log.Errorf("No recovery plan for AddPiece failing")
+	// todo: cleanup sector / just go retry (requires adding offset param to AddPiece in sector-storage for this to be safe)
+	return nil
+}
+
+func (m *Sealing) AddPieceToAnySector(ctx context.Context, size abi.UnpaddedPieceSize, data storage.Data, deal DealInfo) (abi.SectorNumber, abi.PaddedPieceSize, error) {
+	log.Infof("Adding piece for deal %d (publish msg: %s)", deal.DealID, deal.PublishCid)
+	if (padreader.PaddedSize(uint64(size))) != size {
+		return 0, 0, xerrors.Errorf("cannot allocate unpadded piece")
+	}
+
+	sp, err := m.currentSealProof(ctx)
+	if err != nil {
+		return 0, 0, xerrors.Errorf("getting current seal proof type: %w", err)
+	}
+
+	ssize, err := sp.SectorSize()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if size > abi.PaddedPieceSize(ssize).Unpadded() {
+		return 0, 0, xerrors.Errorf("piece cannot fit into a sector")
+	}
+
+	if deal.PublishCid == nil {
+		return 0, 0, xerrors.Errorf("piece must have a PublishCID")
+	}
+
+	m.inputLk.Lock()
+	if _, exist := m.pendingPieces[*deal.PublishCid]; exist {
```

Comment:
Doing this by publishcid is very broken with multiple deals per message, needs fixing